File: 775-n-ary-tree-preorder-traversal/n-ary-tree-preorder-traversal.py
"""
# Definition for a Node.
class Node:
    def __init__(self, val=None, children=None):
        self.val = val
        self.children = children
"""

class Solution:
    def preorder(self, root: 'Node') -> List[int]:
        # def dfs(node):
        #     if not node:
        #         return
        #     result.append(node.val)
        #     for child in node.children:
        #         dfs(child)
        # result = []
        # dfs(root)
        # return result

        if not root:
            return []
        stack, result = [root], []
        while stack:
            node = stack.pop()
            result.append(node.val)
            stack.extend(reversed(node.children))
        return result

File: 775-n-ary-tree-preorder-traversal/n-ary-tree-preorder-traversal.js
/**
 * // Definition for a _Node.
 * function _Node(val, children) {
 *    this.val = val;
 *    this.children = children;
 * };
 */

/**
 * @param {_Node|null} root
 * @return {number[]}
 */
var preorder = function(root) {
    /* const result = [];

    function dfs(node) {
        if (!node) return;

        result.push(node.val);
        for (let child of node.children) {
            dfs(child);
        }
    }
    dfs(root);
    return result; */
    if (!root) return [];
    const stack = [root];
    const result = [];
    while (stack.length !== 0) {
        let node = stack.pop();
        result.push(node.val);
        for (let i = node.children.length - 1; i >= 0; i--) {
            stack.push(node.children[i]);
        }
    }
    return result;
};